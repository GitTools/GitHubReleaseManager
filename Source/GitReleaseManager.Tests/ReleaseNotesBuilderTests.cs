//-----------------------------------------------------------------------
// <copyright file="ReleaseNotesBuilderTests.cs" company="GitTools Contributors">
//     Copyright (c) 2015 - Present - GitTools Contributors
// </copyright>
//-----------------------------------------------------------------------

namespace GitReleaseManager.Tests
{
    using System;
    using System.Linq;
    using ApprovalTests;
    using GitReleaseManager.Core;
    using GitReleaseManager.Core.Configuration;
    using GitReleaseManager.Core.Helpers;
    using GitReleaseManager.Core.Model;
    using NUnit.Framework;

    [TestFixture]
    public class ReleaseNotesBuilderTests
    {
        [Test]
        public void NoCommitsNoIssues()
        {
            var exception = Assert.Throws<AggregateException>(() => AcceptTest(0));
            Assert.That(exception.InnerException, Is.Not.Null.And.TypeOf<InvalidOperationException>());
        }

        [Test]
        public void NoCommitsSomeIssues()
        {
            AcceptTest(0, CreateIssue(1, "Bug"), CreateIssue(2, "Feature"), CreateIssue(3, "Improvement"));
            Assert.True(true); // Just to make sonarlint happy
        }

        [Test]
        public void SomeCommitsNoIssues()
        {
            var exception = Assert.Throws<AggregateException>(() => AcceptTest(5));
            Assert.That(exception.InnerException, Is.Not.Null.And.TypeOf<InvalidOperationException>());
        }

        [Test]
        public void SomeCommitsSomeIssues()
        {
            AcceptTest(5, CreateIssue(1, "Bug"), CreateIssue(2, "Feature"), CreateIssue(3, "Improvement"));
            Assert.True(true); // Just to make sonarlint happy
        }

        [Test]
        public void SingularCommitsNoIssues()
        {
            var exception = Assert.Throws<AggregateException>(() => AcceptTest(1));
            Assert.That(exception.InnerException, Is.Not.Null.And.TypeOf<InvalidOperationException>());
        }

        [Test]
        public void SingularCommitsSomeIssues()
        {
            AcceptTest(1, CreateIssue(1, "Bug"), CreateIssue(2, "Feature"), CreateIssue(3, "Improvement"));
            Assert.True(true); // Just to make sonarlint happy
        }

        [Test]
        public void SingularCommitsSingularIssues()
        {
            AcceptTest(1, CreateIssue(1, "Bug"));
            Assert.True(true); // Just to make sonarlint happy
        }

        [Test]
        public void NoCommitsSingularIssues()
        {
            AcceptTest(0, CreateIssue(1, "Bug"));
            Assert.True(true); // Just to make sonarlint happy
        }

        [Test]
        public void SomeCommitsSingularIssues()
        {
            AcceptTest(5, CreateIssue(1, "Bug"));
            Assert.True(true); // Just to make sonarlint happy
        }

        [Test]
        public void SingularCommitsWithHeaderLabelAlias()
        {
            var config = new Config();
            config.LabelAliases.Add(new LabelAlias
            {
                Name = "Bug",
                Header = "Foo",
            });

            AcceptTest(1, config, CreateIssue(1, "Bug"));
            Assert.True(true); // Just to make sonarlint happy
        }

        [Test]
        public void SomeCommitsWithPluralizedLabelAlias()
        {
            var config = new Config();
            config.LabelAliases.Add(new LabelAlias
            {
                Name = "Help Wanted",
                Plural = "Bar",
            });

            AcceptTest(5, config, CreateIssue(1, "Help Wanted"), CreateIssue(2, "Help Wanted"));
            Assert.True(true); // Just to make sonarlint happy
        }

        [Test]
        public void SomeCommitsWithoutPluralizedLabelAlias()
        {
            AcceptTest(5, CreateIssue(1, "Help Wanted"), CreateIssue(2, "Help Wanted"));
            Assert.True(true); // Just to make sonarlint happy
        }

        [Test]
        public void NoCommitsWrongIssueLabel()
        {
            var exception = Assert.Throws<AggregateException>(() => AcceptTest(0, CreateIssue(1, "Test")));
            Assert.That(exception.InnerException, Is.Not.Null.And.TypeOf<InvalidOperationException>());
        }

        [Test]
        public void SomeCommitsWrongIssueLabel()
        {
            var exception = Assert.Throws<AggregateException>(() => AcceptTest(5, CreateIssue(1, "Test")));
            Assert.That(exception.InnerException, Is.Not.Null.And.TypeOf<InvalidOperationException>());
        }

        [Test]
        public void CorrectlyExcludeIssues()
        {
            AcceptTest(5, CreateIssue(1, "Build"), CreateIssue(2, "Bug"));
            Assert.True(true); // Just to make sonarlint happy
        }

        private static void AcceptTest(int commits, params Issue[] issues)
        {
            AcceptTest(commits, null, issues);
            Assert.True(true); // Just to make sonarlint happy
        }

        private static void AcceptTest(int commits, Config config, params Issue[] issues)
        {
            var fakeClient = new FakeGitHubClient();
            var fileSystem = new FileSystem();
            var currentDirectory = Environment.CurrentDirectory;
            var configuration = config ?? ConfigurationProvider.Provide(currentDirectory, fileSystem);

            fakeClient.Milestones.Add(CreateMilestone("1.2.3"));

            fakeClient.NumberOfCommits = commits;

            foreach (var issue in issues)
            {
                fakeClient.Issues.Add(issue);
            }

            var builder = new ReleaseNotesBuilder(fakeClient, "TestUser", "FakeRepository", "1.2.3", configuration);
            var notes = builder.BuildReleaseNotes().Result;

            Approvals.Verify(notes);
        }

        private static Milestone CreateMilestone(string version)
        {
            return new Milestone
            {
                Title = version,
                HtmlUrl = "https://github.com/gep13/FakeRepository/issues?q=milestone%3A" + version,
                Version = new Version(version),
            };
        }

        private static Issue CreateIssue(int number, params string[] labels)
        {
            return new Issue
            {
                Number = number.ToString(),
                Labels = labels.Select(l => new Label { Name = l }).ToList(),
                HtmlUrl = "http://example.com/" + number,
                Title = "Issue " + number,
            };
        }
    }
}