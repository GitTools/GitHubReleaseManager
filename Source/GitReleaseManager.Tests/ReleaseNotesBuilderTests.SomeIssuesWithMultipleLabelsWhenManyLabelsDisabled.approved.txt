As part of this release we had [5 commits](https://github.com/TestUser/FakeRepository/commits/1.2.3) which resulted in [2 issues](https://github.com/FakeRepository/issues/issues?milestone=0&state=closed) being closed.


__Help Wanteds__

- [__#1__](http://example.com/1) Issue 1
- [__#2__](http://example.com/2) Issue 2

 